plugins/redfish/fu-plugin-redfish.c

```diff
@@ -414,6 +418,70 @@ fu_plugin_redfish_startup(FuPlugin *plugin, GError **error)
 	return fu_backend_setup(FU_BACKEND(data->backend), error);
 }
 
+static gboolean
+fu_plugin_redfish_cleanup_cb(FuDevice *device, gpointer user_data, GError **error)
+{
+	FuPlugin *self = FU_PLUGIN(user_data);
+	FuPluginData *data = fu_plugin_get_data(self);
+	return fu_backend_setup(FU_BACKEND(data->backend), error);
+}
+
+static gboolean
+fu_plugin_redfish_cleanup(FuPlugin *self,
+			  FuDevice *device,
+			  FuProgress *progress,
+			  FwupdInstallFlags flags,
+			  GError **error)
+{
+	FuPluginData *data = fu_plugin_get_data(self);
+	g_autoptr(FuRedfishRequest) request = fu_redfish_backend_request_new(data->backend);
+	g_autoptr(JsonBuilder) builder = json_builder_new();
+	g_autoptr(GPtrArray) devices = NULL;
+
+	/* nothing to do */
+	if (!fu_device_has_private_flag(device, FU_REDFISH_DEVICE_FLAG_MANAGER_RESET))
+		return TRUE;
+
+	/* ask the BMC to reboot */
+	json_builder_begin_object(builder);
+	json_builder_set_member_name(builder, "ResetType");
+	json_builder_add_string_value(builder, "ForceRestart");
+	json_builder_end_object(builder);
+	if (!fu_redfish_request_perform_full(request,
+					     "/redfish/v1/Managers/1/Actions/Manager.Reset",
+					     "POST",
+					     builder,
+					     FU_REDFISH_REQUEST_PERFORM_FLAG_NONE,
+					     error)) {
+		g_prefix_error(error, "failed to reset manager: ");
+		return FALSE;
+	}
+
+	/* remove all the devices */
+	devices = fu_backend_get_devices(FU_BACKEND(data->backend));
+	for (guint i = 0; i < devices->len; i++) {
+		FuDevice *device_tmp = g_ptr_array_index(devices, i);
+		fu_backend_device_removed(FU_BACKEND(data->backend), device_tmp);
+	}
+
+	/* wait for the BMC to come back */
+	fu_backend_invalidate(FU_BACKEND(data->backend));
+	g_usleep(10 * 1000);
+	if (!fu_device_retry_full(device,
+				  fu_plugin_redfish_cleanup_cb,
+				  FU_REDFISH_PLUGIN_CLEANUP_RETRIES_COUNT,
+				  FU_REDFISH_PLUGIN_CLEANUP_RETRIES_DELAY,
+				  self,
+				  error)) {
+		g_prefix_error(error, "manager failed to come back: ");
+		return FALSE;
+	}
+
+	/* work around an XCC bug... */
+	g_usleep(40 * G_USEC_PER_SEC);
+	return fu_plugin_redfish_coldplug(self, error);
```

Comment:
Why shouldn't this be a detach ? Seems a long time for a plugin to hang.

Reply:
Attach is supposed to put one device into a specific mode, pushing it out of programmer mode or locking it again in some way. The ->detach() and ->attach() vfuncs are only called on the specific device we're updating, and there's no provision for calling it on another device. For instance it might be valid to want to restart the BMC after you've installed a power supply update for instance -- or to restart the mouse main firmware when installing a mouse radio update. For this I think the composite (all plugins) ->cleanup() is the only thing that makes sense. I do agree we need to use the now-passed-in FuProgress to actually set up the timeouts and provide some feedback to the user.

Reply:
I think the source of my confusion is this is run from **plugin** cleanup not **composite** cleanup.  Thanks for explaining.